Add class and module name completion

laginha87 opened the pull request:
Hi all
This adds autocomplete for class names and module names.
It doesn't show any documentation for it yet.
Will add a gif to demo.

faustinoaq commented:
![image](https://user-images.githubusercontent.com/3067335/39657323-6ccb0710-4fcc-11e8-8b25-2fb403f28a41.png)

Works like a charm! :tada:

bew commented on spec/scry/completion_provider_spec.cr:
```diff
@@ -128,9 +128,17 @@ module Scry
+      it_completes("A", %w(Array Atomic ArgumentError AtExitHandlers))
+      it_completes("JSON::P", %w(ParseException Parser PullParser))
+      it_completes("JSON::Pa", %w(ParseException Parser))
+      it_completes("JSO", %w(JSON JSON::Any JSON::Builder JSON::Error JSON::Lexer JSON::Lexer::IOBased JSON::Lexer::StringBased JSON::ParseException JSON::Parser JSON::PullParser JSON::Token))
```
Can you add a spec for completion after `JSON::` ?
And after `::` ? (if it works)

bew commented on src/scry/completion/method_db.cr:
```diff
@@ -20,6 +20,10 @@ module Scry::Completion
     end
 
+    def type_match(text : String) : Array(String)
+      @db.keys.select(&.starts_with?(text)).reject(&.ends_with?(".class"))
```
`select!`
`reject!`

faustinoaq replied:
@bew Why? are mutable methods faster?

bew replied:
Yeah I think so, they don't allocate a new array each time.

bew commented on src/scry/completion/method_db.cr:
```diff
@@ -53,39 +57,54 @@ module Scry::Completion
 
   class Visitor < Crystal::Visitor
     property classes
+    property class_name = ""
```
I think it's better to use `nil` to denote the absence of `class_name`, so the compiler can help you instead of doing empty checks.

laginha87 replied:
I think it's easier with an empty string cause we only interpolate stuff, otherwise I'll have to use not_nil! calls

laginha87 replied:
@class_name is assigned  the value of the generate_classname which will be an empty string if the queues are empty.

faustinoaq replied:
```crystal
property class_name : String?
```

and

```crystal
if this_class_name = class_name
  # Use this_cass_name here...
end
```

WDYT?

faustinoaq replied:
Although I see no issues with initializing the `class_name = ""` 

@bew Is this a bad practice or something? :sweat_smile:

bew replied:
I'll simply say:

```
method_receiver = node.receiver ? "#{@class_name}.class" : @class_name
```
In this case, does it make sense (allowed) to have `@class_name` as an empty string?
I don't think so, so it must be made difficult/impossible to do.

laginha87 replied:
that won't happen cause if class_name is "" then class_queue is empty and in here we opt out https://github.com/crystal-lang-tools/scry/pull/100/files/d05348c07392320c379be63575985d0f83398c56#diff-45d877a34e1040cf2af4c4c46cdcf290R73
The idea of the classname is just to cache the output of generate_class_name which never returns nil, so initialising it as nil doesn't seem like the right way to handle it.

faustinoaq replied:
@laginha87 Oh, ok, I think `class_name = ""` is not that bad @bew WDYT?

bew replied:
yeah let's not block the PR for a tiny non-straight-forward thing like that, it's ok

bew commented on src/scry/completion/method_db.cr:
```diff
@@ -53,39 +57,54 @@ module Scry::Completion
-      @classes[node.name.to_s] = [] of MethodDBEntry
-      @class_queue << node.name.to_s
+      @class_queue << node.name
+      @class_name = generate_name(@module_queue + @class_queue)
```
Does this still work if you have a module inside a class inside a module?

Why are you separating module & classes queues?

laginha87 replied:
oh, I didn't even know that could happen

laginha87 replied:
It's seperate cause we only store methods for class, for now.

laginha87 replied:
@bew I tried to change this to use classes and modules in the same queue but started getting some tests failing not really sure why. I cut a new branch to add this and module functions as well.

bew replied:
can you create a new issue to track and not forget this? It's quite important
thanks

bew replied:
Also, include what you tried and your results, it might help find the actual problem ;)

bew commented on spec/scry/completion_provider_spec.cr:
```diff
@@ -136,8 +136,11 @@ module Scry
       _kind_ = CompletionItemKind::Class
       it_completes("A", %w(Array Atomic ArgumentError AtExitHandlers))
+      it_completes(" ::A", %w(Array Atomic ArgumentError AtExitHandlers))
+      it_completes(" ::", %w())
```
why nothing? It should give every top level classes & constants

faustinoaq replied:
@bew I think listing all classes would be very slow :sweat_smile: , no?

By example on Workspace symbols, if query is empty, you get this:

![screenshot_20180509_193607](https://user-images.githubusercontent.com/3067335/39846247-3932427c-53c0-11e8-8d8a-7f8457dfea51.png)

bew replied:
@faustinoaq you're showing symbols here, not completion.

Slow? I don't know. 

So for now it only gives an empty list?

laginha87 replied:
I couldn't figure out why it didn't return anything :(

faustinoaq replied:
No problem, I think is ok :+1:

bew replied:
@laginha87 let's not block this pr as it's not so important, but please open an issue to keep track of `::` completion, and fix it one day, or decide what we do about it

laginha87 replied:
I actually wanna tackle doing a context aware comlpetion for this at some point .
for example:
```
module A
     ::
```
should complete with modules and classes inside A as well

laginha87 replied:
@bew added https://github.com/crystal-lang-tools/scry/issues/118

bew replied:
@laginha87 there is no need for contextual completion, `::` completions will ALWAYS be the top level things, that's what it exists for, there is no contextual things going on for types that starts with `::`

faustinoaq commented:
@crystal-lang-tools/scry Can we get another review here?

laginha87 commented:
@bew, thanks for reviewing the code, can you approve this or is there something I missed?

bew commented on spec/scry/completion_provider_spec.cr:
```diff
@@ -128,9 +128,20 @@ module Scry
       end
     end
+
+    context "classname completion" do
```
Note: it can also complete module names (and structs), not only class names, maybe change the spec name?

bew commented on spec/scry/completion_provider_spec.cr:
```diff
@@ -128,9 +128,20 @@ module Scry
+
+    context "classname completion" do
+      _kind_ = CompletionItemKind::Class
+      it_completes("A", %w(Array Atomic ArgumentError AtExitHandlers))
```
This means that this spec rely on external code... A change to the stdlib, an addition/removal of a class starting with `A` will break this spec...

There are other problems with how the specs currently works, like:
- You can't add a description of what that spec is actually spec-ing
- They rely on external code (see above)
- They are not self-contained: by just reading the spec you don't directly have all the data to understand it (you have to always lookup other files, AND they are shared between many specs)

This should probably go to a new issue though. _I'm just too tired right now_ ^^

laginha87 replied:
1. The test will always break with changes to stdlib, we are testing the autocomplete on it I don't think it's something we should worry about.
2. What other files do you have to check to understand the specs? My idea for this file was to have tests that almost did an end-to-end test on the completion features and not dependant on the actual implementation.

bew commented on src/scry/completion/class_module_context.cr:
```diff
@@ -0,0 +1,22 @@
+    def find
+      target_header_index = @target.rindex(":")
+      @method_db.type_match(@target).map do |label|
+        label = label[(target_header_index + 1)..-1] if target_header_index
```
just a style comment: At first I didn't saw the `if` at the end of the line, and was wondering 'how the hell does that compile..."

I think that line is too 'complicated' / 'long' for a suffix `if`, and would suggest to make it a multi-line `if`.

bew commented on src/scry/completion/class_module_context.cr:
```diff
@@ -0,0 +1,22 @@
+          label: label,
+          insert_text: label,
+          kind: CompletionItemKind::Class,
+          detail: label,
```
would it be better to have the full path (e.g: `JSON::Any`, not just `Any`) in details? (can someone test this, see how it goes?)

laginha87 replied:
you mean in the CompletionItem class ?

bew commented on src/scry/completion/method_db.cr:
```diff
@@ -20,6 +20,10 @@ module Scry::Completion
     end
 
+    def type_match(text : String) : Array(String)
+      @db.keys.select!(&.starts_with?(text)).reject!(&.ends_with?(".class"))
```
```cr
@db.keys.select! { |k| k.starts_with?(text) && !k.ends_with?(".class") }
```
? (less allocations / simpler logic / ...)

laginha87 commented:
@bew thanks for the review 👍 
I addressed most of the issues you pointed out.
What you brought up about the test is valid but think it should be taken care of in another PR.
Didn't understand what you meant by "would it be better to have the full path (e.g: JSON::Any, not just Any) in details? (can someone test this, see how it goes?)" 
Can you take another look and maybe approve 😅?

bew commented:
Actually it's not important

bew commented on spec/fixtures/completion/sample.cr:
```diff
@@ -1,6 +1,6 @@
 require "./tree"
 
-class A
+class B
```
By "looking another file to understand the specs" I meant this file!

faustinoaq commented:
@laginha87 Thank you! :tada:
